Indicators/AutoRegressiveIntegratedMovingAverage.cs

```diff
@@ -0,0 +1,300 @@
+                errAr += Math.Pow(residual, 2);
+                _residuals.Add(residual);
+            }
+
+            ArResidualError = errAr / (data.Length - _arOrder - 1);
+            if (_maOrder == 0)
+            {
+                ArParameters = arFits; // Will not be thrown out
+            }
+
+            if (_maOrder > 0) // MA part as in (4) of mbhauser notes.
+            {
+                var appendedData = new List<double[]>();
+                var laggedErrors = LaggedSeries(_maOrder, _residuals.ToArray());
+                for (var i = 0; i < laggedErrors.Length; i++)
+                {
+                    var doubles = lags[i].ToList();
+                    doubles.AddRange(laggedErrors[i]);
+                    appendedData.Add(doubles.ToArray());
+                }
+
+                var maFits = Fit.MultiDim(appendedData.ToArray(), data.Skip(_maOrder).ToArray(),
+                    method: DirectRegressionMethod.NormalEquations, intercept: _intercept);
+                for (var i = _maOrder; i < data.Length; i++) // Calculate the error assoc. with model.
+                {
+                    var paramVector = _intercept
+                        ? Vector.Build.Dense(maFits.Skip(1).ToArray())
+                        : Vector.Build.Dense(maFits);
+                    var residual = data[i] - Vector.Build.Dense(appendedData[i - _maOrder]).DotProduct(paramVector);
+                    errMa += Math.Pow(residual, 2);
+                }
+
+                switch (_intercept)
+                {
+                    case true:
+                        MaResidualError = errMa / (data.Length - Math.Max(_arOrder, _maOrder) - 1);
+                        MaParameters = maFits.Skip(1 + _arOrder).ToArray();
+                        ArParameters = maFits.Skip(1).Take(_arOrder).ToArray();
+                        Intercept = maFits[0];
+                        break;
+                    default:
+                        MaResidualError = errMa / (data.Length - Math.Max(_arOrder, _maOrder) - 1);
+                        MaParameters = maFits.Skip(_arOrder).ToArray();
+                        ArParameters = maFits.Take(_arOrder).ToArray();
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Forecasts the series of the fitted model one point ahead.
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IndicatorDataPoint input)
+        {
+            _rollingData.Add((double) input.Value);
+            if (_rollingData.IsReady)
+            {
+                var arrayData = _rollingData.ToArray();
+                TwoStepFit(arrayData);
+                arrayData = _diffOrder > 0 ? DifferenceSeries(_diffOrder, arrayData) : arrayData;
```

Comment:
`arrayData = _diffOrder > 0 ? DifferenceSeries(_diffOrder, arrayData) : arrayData;` is also done by `TwoStepFit` internally, could we do it once and pass it in?